front-end/src/main/kotlin/org/jetbrains/kotlin/spec/viewer/SpecTestsViewer.kt

```diff
@@ -82,63 +100,92 @@ class SpecTestsViewer {
         val helperFilesPromises = mutableListOf<Promise<Map<String, Any>>>()
 
         helperNames.forEach { helperName ->
-            helperFilesPromises.add(SpecTestsLoader.loadHelperFile(helperName))
+            //todo fix  helperFilesPromises.add(SpecTestsLoader.loadHelperFile(helperName))
         }
 
         return Promise.all(helperFilesPromises.toTypedArray())
     }
 
     fun showViewer(sentenceElement: JQuery) {
-        val tests =
-                sentenceElement.data("tests").unsafeCast<Map<String, Map<String, Map<String, Map<String, String>>>>>()
+        val tests = sentenceElement.data("tests").unsafeCast<Sentence>()
         val sentenceText = "sentence {1}".format(sentenceElement.clone().children(".number-info").text())
 
         currentSentenceTests = tests
         testPopup = Popup(
                 PopupConfig(
                         title = "Test coverage of $sentenceText",
-                        content = TestsCoverageColorsArranger.TEMPLATE,
+                        content = SpecCoverageHighlighter.TEMPLATE,
                         width = 800,
                         height = 300
                 )
         ).apply { open() }
 
         `$`(TEST_AREA_OPTION_SELECTOR).each { _, el ->
             val testArea = `$`(el)
-            if (testArea.attr("value") !in tests) {
+
+
+            if (tests.getTestsByTestArea(TestArea.getByAttribute(testArea.attr("value"))) == null) {
                 testArea.remove()
             }
         }
+
         `$`(TEST_AREA_SELECTOR).`val`(`$`(TEST_AREA_OPTION_SELECTOR).eq(0).`val`().toString())
 
         onTestAreaChange()
     }
 
     fun onTestAreaChange() {
-        val testArea = `$`(TEST_AREA_SELECTOR).`val`().toString().apply { if (isEmpty()) return }
-        val tests = currentSentenceTests[testArea] ?: return
-        val testTypes = mapOf("pos" to "Positive", "neg" to "Negative")
-
-        //TODO: fix appends if test type option is not appended yet
-        tests.keys.reversed().forEach { testType ->
-            `$`(TEST_TYPE_SELECTOR).append(TEST_TYPE_OPTION_TEMPLATE.format(testType, testTypes[testType]
-                    ?: return@forEach))
-        }
+        val testArea = TestArea.getByAttribute(`$`(TEST_AREA_SELECTOR).`val`().toString().apply { if (isEmpty()) return })
+        currentSentenceTests.getTestsByTestArea(testArea) ?: return
+        addTestTypeOption(testArea, TestType.Positive)
+        addTestTypeOption(testArea, TestType.Negative)
 
         `$`(TEST_TYPE_SELECTOR).show().`val`(`$`(TEST_TYPE_OPTION_SELECTOR).eq(0).`val`().toString())
 
         onTestTypeChange()
     }
 
+    private fun addTestTypeOption(testArea: TestArea, testType: TestType) {
```

Comment:
What is the " test type option"?

Reply:
hm, actually it's better to remove this function, so I've done